README.rst

```diff
@@ -74,11 +68,167 @@ To install in a virtual environment in your current project:
+must end with ``/`` when provided as a full path.
+
+All numbers are unsigned.
+
+All values are aligned with respect to the start of the packet.
+
+Status bytes are ``0x01`` for OK and ``0x02`` for error. Other values for error may be used for specific commands.
+
+``0x10`` - Read a file
+++++++++++++++++++++++
+
+Given a full path, returns the full contents of the file.
+
+The header is four fixed entries and a variable length path:
+
+* Command: Single byte. Always ``0x10``.
+* 1 Byte reserved for padding.
+* Path length: 16-bit number encoding the encoded length of the path string.
+* Chunk offset: 32-bit number encoding the offset into the file to start the first chunk.
+* Chunk size: 32-bit number encoding the amount of data that the client can handle in the first reply.
+* Path: UTF-8 encoded string that is *not* null terminated. (We send the length instead.)
+
+The server will respond with:
+* Command: Single byte. Always ``0x11``.
+* Status: Single byte.
+* 2 Bytes reserved for padding.
+* Chunk offset: 32-bit number encoding the offset into the file of this chunk.
+* Total length: 32-bit number encoding the total file length.
+* Chunk length: 32-bit number encoding the length of the read data up to the chunk size provided in the header.
+* Chunk-length contents of the file starting from the current position.
+
+If the chunk length is smaller than the total length, then the client will request more data by sending:
+* Command: Single byte. Always ``0x12``.
+* Status: Single byte. Always OK for now.
+* 2 Bytes reserved for padding.
+* Chunk offset: 32-bit number encoding the offset into the file to start the next chunk.
+* Chunk size: 32-bit number encoding the number of bytes to read. May be different than the original size. Does not need to be limited by the total size.
+
+The transaction is complete after the server has replied with all data. (No acknowledgement needed from the client.)
+
+``0x20`` - Write a file
++++++++++++++++++++++++
+
+Writes the content to the given full path. If the file exists, it will be overwritten. Content may be written as received so an interrupted transfer may lead to a truncated file.
+
+Offset larger than the existing file size will introduce zeros into the gap.
+
+The header is four fixed entries and a variable length path:
+
+* Command: Single byte. Always ``0x20``.
+* 1 Byte reserved for padding.
+* Path length: 16-bit number encoding the encoded length of the path string.
+* Offset: 32-bit number encoding the starting offset to write.
+* Total size: 32-bit number encoding the total length of the file contents.
+* Path: UTF-8 encoded string that is *not* null terminated. (We send the length instead.)
+
+The server will repeatedly respond until the total length has been transferred with:
+* Command: Single byte. Always ``0x21``.
+* Status: Single byte. ``0x01`` if OK. ``0x02`` if any parent directory is missing or a file.
```

Comment:
I think we should also have `busy` status, since BLE can have concurrent connection with multiple clients. Busy would prevent 2 clients trying to write to a same file.

For concurrent connection, we should also find a sensible way to response to Read command while an file is writing to it since it could change the file size.